lib/puppet/provider/exec/powershell.rb

```diff
@@ -25,17 +26,55 @@
         }
   EOT
 
+  def self.upgrade_message
+    Puppet.warning <<-UPGRADE
+The current Puppet version is outdated and uses a library that was
+previously necessary on the current Ruby verison to support a colored console.
+
+Unfortunately this library prevents the PowerShell module from using a shared
+PowerShell process to dramatically improve the performance of resource
+application.
+
+To enable these improvements, it is suggested to upgrade to any x64 version of
+Puppet (including 3.x), or to a Puppet version newer than 3.x.
+    UPGRADE
+  end
+
+  if Facter.value(:uses_win32console)
+    upgrade_message
+  end
+
+  def self.powershell_args
+    ps_args = ['-NoProfile', '-NonInteractive', '-Sta', '-NoLogo', '-ExecutionPolicy', 'Bypass', '-Command']
+    ps_args << '-' if !Facter.value(:uses_win32console)
+    ps_args
+  end
+
+  def ps_manager
+    PuppetX::PowerShell::PowerShellManager.instance("#{command(:powershell)} #{self.class.powershell_args.join(' ')}")
+  end
+
   def run(command, check = false)
-    write_script(command) do |native_path|
-      # Ideally, we could keep a handle open on the temp file in this
-      # process (to prevent TOCTOU attacks), and execute powershell
-      # with -File <path>. But powershell complains that it can't open
-      # the file for exclusive access. If we close the handle, then an
-      # attacker could modify the file before we invoke powershell. So
-      # we redirect powershell's stdin to read from the file. Current
-      # versions of Windows use per-user temp directories with strong
-      # permissions, but I'd rather not make (poor) assumptions.
-      return super("cmd.exe /c \"\"#{native_path(command(:powershell))}\" #{args} -Command - < \"#{native_path}\"\"", check)
+    feature_flag = false
+    if feature_flag
+      write_script(command) do |native_path|
+        # Ideally, we could keep a handle open on the temp file in this
+        # process (to prevent TOCTOU attacks), and execute powershell
+        # with -File <path>. But powershell complains that it can't open
+        # the file for exclusive access. If we close the handle, then an
+        # attacker could modify the file before we invoke powershell. So
+        # we redirect powershell's stdin to read from the file. Current
+        # versions of Windows use per-user temp directories with strong
+        # permissions, but I'd rather not make (poor) assumptions.
+        return super("cmd.exe /c \"\"#{native_path(command(:powershell))}\" #{args} -Command - < \"#{native_path}\"\"", check)
```

Comment:
Writing the file is still OK as we do it safely.  As it turns out, it also appears necessary to allow the user to continue to use `exit` in their scripts to generate an exit code.

To spell it out a little more, doing the following will exit the `Powershell` instance, and there's no way for us to capture the exit code at this point:

```powershell
$ps.AddScript('exit 1')
$output = $ps.Invoke()
```

However, given a file `foo.ps1`

```powershell
exit 1
```

And an execution like:

```powershell
$ps.AddScript('./foo.ps1')
$output = $ps.Invoke()

$ps.AddScript('echo $LASTEXITCODE')
$exitcode = $ps.Invoke()
```

We can actually get to the exit code.